// rumia/dist/res/connect.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocket = void 0;
var ws_1 = __importDefault(require("ws"));
var reconnecting_websocket_1 = __importDefault(require("reconnecting-websocket"));
var WebSocket = /** @class */ (function () {
    function WebSocket(url, maxRetries) {
        var _this = this;
        this.socket = new reconnecting_websocket_1.default(url, [], {
            WebSocket: ws_1.default,
            maxRetries: maxRetries
        });
        this.onopen = null;
        this.onmessage = null;
        this.socket.onopen = this.onopen;
        this.socket.onmessage = function (event) {
            if (_this.onmessage) {
                try {
                    _this.onmessage(JSON.parse(event.data));
                }
                catch (error) {
                    _this.onmessage(event.data);
                }
            }
        };
    }
    WebSocket.prototype.send = function (data) {
        this.socket.send(JSON.stringify(data));
    };
    return WebSocket;
}());
exports.WebSocket = WebSocket;
